### PR TITLE
[EE-92][product-service] Create Brand API.

### PR DESCRIPTION
**Pull request by akhter-rakib**
-- added BrandDto and Brand entity
-- added BrandMapper
-- added BrandController, BrandService and BrandRepository
-- added @MappedSuperclass to fix the baseEntity not added   into every entity
-- added dependency for validation
-- validate the BrandDto

### CODE REVIEW

File: product-service/src/main/java/com/extremecoder/productservice/controller/BrandController.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import javax.validation.Valid;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/v1/brands")
+public class BrandController {
```
Comment (shshetudev) [inline product-service/src/main/java/com/extremecoder/productservice/controller/BrandController.java:21]: log missing. Add @Slf4j

File: product-service/src/main/java/com/extremecoder/productservice/controller/BrandController.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    private final BrandService brandService;
+
+    @PostMapping
+    public Response saveBrand(@Valid @RequestBody BrandDto brandDto) {
```
Comment (shshetudev) [inline product-service/src/main/java/com/extremecoder/productservice/controller/BrandController.java:25]: Add Response<BrandDto>

File: product-service/src/main/java/com/extremecoder/productservice/controller/BrandController.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        brandService.save(brandDto);
+        return ResponseBuilder.getSuccessResponse(HttpStatus.OK, "Brand Save Successfully", "");
```
Comment (shshetudev) [inline product-service/src/main/java/com/extremecoder/productservice/controller/BrandController.java]: 2nd parameter, String.join("Brand {} saved successfully", brandDto.getName())
3rd parameter, brandDto

File: product-service/src/main/java/com/extremecoder/productservice/model/Brand.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+
+@Setter
+@Getter
+@Entity
+public class Brand extends BaseEntity {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
Comment (shshetudev) [inline product-service/src/main/java/com/extremecoder/productservice/model/Brand.java]: Use static import

File: product-service/src/main/java/com/extremecoder/productservice/service/impl/BrandServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+@Service
+@RequiredArgsConstructor
+@Slf4j
+public class BrandServiceImpl implements BrandService {
+    private final BrandRepository brandRepository;
+    private final BrandMapper brandMapper;
+
+    @Override
+    public void save(BrandDto brandDto) {
+        Brand brand = MappingProvider.BRAND_MAPPER.toEntity(brandDto);
```
Comment (shshetudev) [inline product-service/src/main/java/com/extremecoder/productservice/service/impl/BrandServiceImpl.java]: 1. Use static import `toEntity(brandDto)` for this method.
2. Use like this:
`return brandRepository.save(BRAND_MAPPER.toEntity(brandDto))`
3. `Saved brand with id {}`

File: product-service/src/main/java/com/extremecoder/productservice/model/Brand.java (hunk 4 above)
Comment (shshetudev) [inline product-service/src/main/java/com/extremecoder/productservice/model/Brand.java]: static import `GenerationType.IDENTITY`

File: product-service/src/main/java/com/extremecoder/productservice/mapper/BrandMapper.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
+package com.extremecoder.productservice.mapper;
+
+import com.extremecoder.productservice.dto.BrandDto;
+import com.extremecoder.productservice.model.Brand;
+import org.mapstruct.Mapper;
+
+@Mapper(componentModel = "spring")
```
Comment (shshetudev) [inline product-service/src/main/java/com/extremecoder/productservice/mapper/BrandMapper.java]: Remove `componentModel = "spring"`

File: product-service/src/main/java/com/extremecoder/productservice/controller/BrandController.java (hunk 2 above)
Comment (shshetudev) [reply product-service/src/main/java/com/extremecoder/productservice/controller/BrandController.java:25]: Add `log.info("Saving brand with name {}", brandDto.getName())`

File: product-service/src/main/java/com/extremecoder/productservice/mapper/BrandMapper.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+public interface BrandMapper {
+
+    Brand toEntity(BrandDto brandDto);
+
+    BrandDto toDto(Brand brand);
```
Comment (shshetudev) [inline product-service/src/main/java/com/extremecoder/productservice/mapper/BrandMapper.java:14]: Add 
```
List<BrandDto> toDtos(List<Brand> brands);
List<Brand> toEntites(List<BrandDto> brandDtos);
```

File: product-service/src/main/java/com/extremecoder/productservice/model/BaseEntity.java [hunk 8]
```diff
@@ -2,9 +2,11 @@
...
 import com.extremecoder.productservice.enums.ActiveStatus;
 
+import javax.persistence.MappedSuperclass;
 import java.io.Serializable;
 import java.util.Date;
 
+@MappedSuperclass
 public class BaseEntity implements Serializable {
 
     private ActiveStatus activeStatus = ActiveStatus.ACTIVE;
```
Comment (shshetudev) [inline product-service/src/main/java/com/extremecoder/productservice/model/BaseEntity.java]: static import

File: product-service/src/main/java/com/extremecoder/productservice/repository/BrandRepository.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package com.extremecoder.productservice.repository;
+
+import com.extremecoder.productservice.enums.ActiveStatus;
+import com.extremecoder.productservice.model.Brand;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface BrandRepository extends JpaRepository<Brand, Long> {
+    @Override
+    default void delete(Brand entity) {
+        entity.setActiveStatus(ActiveStatus.DELETED);
```
Comment (shshetudev) [inline product-service/src/main/java/com/extremecoder/productservice/repository/BrandRepository.java]: static import

**Review by shshetudev — COMMENTED** [review]
Great job bro! Appreciate your effort!

File: product-service/src/main/java/com/extremecoder/productservice/service/impl/BrandServiceImpl.java (hunk 5 above)
Comment (akhter-rakib) [reply product-service/src/main/java/com/extremecoder/productservice/service/impl/BrandServiceImpl.java]: Done

File: product-service/src/main/java/com/extremecoder/productservice/model/Brand.java (hunk 4 above)
Comment (akhter-rakib) [reply product-service/src/main/java/com/extremecoder/productservice/model/Brand.java]: Done

Comment (akhter-rakib) [reply product-service/src/main/java/com/extremecoder/productservice/model/Brand.java]: Done

File: product-service/src/main/java/com/extremecoder/productservice/repository/BrandRepository.java (hunk 9 above)
Comment (akhter-rakib) [reply product-service/src/main/java/com/extremecoder/productservice/repository/BrandRepository.java]: Done

File: product-service/src/main/java/com/extremecoder/productservice/model/BaseEntity.java (hunk 8 above)
Comment (akhter-rakib) [reply product-service/src/main/java/com/extremecoder/productservice/model/BaseEntity.java]: Done

File: product-service/src/main/java/com/extremecoder/productservice/mapper/BrandMapper.java (hunk 7 above)
Comment (akhter-rakib) [reply product-service/src/main/java/com/extremecoder/productservice/mapper/BrandMapper.java:14]: Added

File: product-service/src/main/java/com/extremecoder/productservice/mapper/BrandMapper.java (hunk 6 above)
Comment (akhter-rakib) [reply product-service/src/main/java/com/extremecoder/productservice/mapper/BrandMapper.java]: Removed

File: product-service/src/main/java/com/extremecoder/productservice/controller/BrandController.java (hunk 1 above)
Comment (akhter-rakib) [reply product-service/src/main/java/com/extremecoder/productservice/controller/BrandController.java:21]: Added

File: product-service/src/main/java/com/extremecoder/productservice/controller/BrandController.java (hunk 2 above)
Comment (akhter-rakib) [reply product-service/src/main/java/com/extremecoder/productservice/controller/BrandController.java:25]: Added

**Pull request merged**